redeem/Pipe.py

```diff
@@ -29,45 +29,54 @@
 import time
 import os
+import errno
+import termios
 from Gcode import Gcode
 
 
 class Pipe:
-  @staticmethod
-  def check_tty0tty():
-    return (find_executable("tty0tty") is not None)
-
-  @staticmethod
-  def check_socat():
-    return (find_executable("socat") is not None)
-
   def __init__(self, printer, prot):
     self.printer = printer
     self.prot = prot
 
-    pipe_0 = "/dev/" + prot + "_0"
-    pipe_1 = "/dev/" + prot + "_1"
-
-    # Ensure tty0tty is installed and available in the PATH
-    if not Pipe.check_tty0tty() and not Pipe.check_socat():
-      logging.error("Neither tty0tty nor socat found! tty0tty or socat must be installed")
-      raise EnvironmentError("tty0tty and socat not found")
-
-    if Pipe.check_tty0tty():
-      p = subprocess.Popen(["tty0tty", pipe_0, pipe_1], stderr=subprocess.PIPE)
-      p.stderr.readline()
-
-    elif Pipe.check_socat():
-      p = subprocess.Popen([
-          "socat", "-d", "-d", "-lf", "/var/log/redeem2" + self.prot,
-          "pty,mode=777,raw,echo=0,link=" + pipe_0, "pty,mode=777,raw,echo=0,link=" + pipe_1
-      ],
-                           stderr=subprocess.PIPE)
-      while not os.path.exists(pipe_0):
-        time.sleep(0.1)
-    self.rd = open(pipe_0, "r")
-    self.wr = os.open(pipe_0, os.O_WRONLY)
-    logging.info("Pipe " + self.prot + " open. Use '" + pipe_1 + "' to " "communicate with it")
+    (master_fd, slave_fd) = os.openpty()
+    slave = os.ttyname(slave_fd)
+
+    # switch to "raw" mode - these constants come from the manpage for termios under cfmakeraw()
+    master_attr = termios.tcgetattr(master_fd)
+    master_attr[0] &= ~(termios.IGNBRK | termios.BRKINT | termios.PARMRK | termios.ISTRIP
+                        | termios.INLCR | termios.IGNCR | termios.ICRNL | termios.IXON)
+    master_attr[1] &= ~termios.OPOST
+    master_attr[2] &= ~(termios.CSIZE | termios.PARENB)
+    master_attr[3] &= ~(termios.ECHO | termios.ECHONL | termios.ICANON | termios.ISIG
+                        | termios.IEXTEN)
+    master_attr[3] |= termios.CS8
+    termios.tcsetattr(master_fd, termios.TCSADRAIN, master_attr)
+
+    # Fun detail: master will always show as /dev/ptmx, but the kernel knows from
+    # the fd which PTY we're using. This means we have to use master_fd instead
+    # of opening master by name.
+
+    logging.info("Opened PTY for " + prot + " and got " + os.ttyname(master_fd) + " and " + slave)
+
+    self.pipe_link = "/dev/" + prot + "_1"
+
+    self.rd = os.fdopen(master_fd, "r")
+    self.wr = os.fdopen(master_fd, "w")
+
+    logging.info("Unlinking " + self.pipe_link)
+    try:
+      os.unlink(self.pipe_link)
+    except OSError, e:
+      # file not found is fine to ignore - anythine else and we should log it
+      if e.errno != errno.ENOENT:
+        logging.error("Failed to unlink " + self.pipe_link + ": " + e.strerror)
+
+    logging.info("re-linking " + self.pipe_link)
+    os.symlink(slave, self.pipe_link)
+    os.chmod(self.pipe_link, 0777)
```

Comment:
Are these the correct permissions?

Reply:
They match the old permissions. I suspect the old permissions were chosen because it was the simplest way to give OctoPrint access.

Reply:
But I think that they are incorrect. 777 gives read, write, and execute to everyone. That is something you might do on a directory.At most, shouldn't the pipe be 666. And, if we can make either the owner or the group "octo", that would be even better.